Fix #2447: Caching should be done on a per-file basis

vojtechjelinek opened the pull request:
1. Hashes of the files are computed and saved.
2. The paths in HTML files are replaced with `filepath + "?v=" + filehash` that means hash is GET request param and is omitted on the backend so we don't need to rename the files.

There are some parts that don't work (the cache won't break with file change):
- Images and directives use `UrlInterpolationService` to create the URL. Also, directive URLs are inside JS files and I currently don't replace URLs in JS files.
- Translates user angular-translate library to load files.

I think the best way to solve those would be to create JSON with hashes and interpolate the URLs on frontend. For translates also create custom file loader, that will use the JSON.

vojtechjelinek committed:
individual file cache slugs

vojtechjelinek committed:
Merged develop into cache-slugs

seanlip commented:
Hi @vojtechjelinek -- I notice that this is not for merge, so just to double check, what would you like me to comment on, exactly?

BenHenning commented:
@vojtechjelinek my main concern with this is it seems to abuse the HTTP query parameters because the parameters aren't being used and that might cause some confusion when people look at this later (especially when debugging). It's not clear that query parameters can even be ignored by app engine. How feasible would it be for us to use a per-file hash solution instead of query parameters?

seanlip commented:
Hi @vojtechjelinek, just FYI, I'm going to pass this to @BenHenning to review, since he's thought about this a lot more than I have. Thanks!

vojtechjelinek commented:
@BenHenning You mean that the hash would be part of the filename? It shouldn't be hard, should I rename the files or modify the handler in app.yaml so it will remove the hash part?

BenHenning commented:
Sorry @vojtechjelinek, yeah that's what I mean (having the hash as part of the filename). Between the two solutions you outlined, what are the pros/cons of each? Which solution do you think is better for us to use?

vojtechjelinek commented:
### Renaming the files
- the whole build process would need refactor
- all files in /build (except for HTML) would have the name with hash inside
### Modifying the handler
- there would be a quite complicated regular expression in app.yaml that would remove the hash part
- all files in /build would have the name without hash

I think modifying the handler is better, but it could be quite hard to understand the regular expression that will transform the requests.

seanlip commented:
Hi @vojtechjelinek, I'm not quite sure what you mean when talking about the regex in the second solution. Please could you elaborate? Also, how would this work in dev?

BenHenning commented:
Thanks @vojtechjelinek. Could you specifically call out the upsides and
downsides of each approach, including details about dev vs prod
differences? Can you be more clear on how much of the build system would
need to change for the file name approach? How much dev time would each one
take? What are the risks? Do we need to change our release process? Can the
regex approach break in some circumstances?

I think we generally want enough info to be able to make an educated
decision on which approach take. I'm still unsure myself, which is in part
why more information would be helpful. :)

On Thu, May 25, 2017 at 4:14 PM Sean Lip <notifications@github.com> wrote:

> Hi @vojtechjelinek <https://github.com/vojtechjelinek>, I'm not quite
> sure what you mean when talking about the regex in the second solution.
> Please could you elaborate? Also, how would this work in dev?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/3454#issuecomment-304149185>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMYdvo_BrowDKwmVqfr72-weBErl_8-Qks5r9gtBgaJpZM4NippJ>
> .
>

vojtechjelinek commented:
Release process wouldn’t need to change in both approaches.
The difference between dev and prod is in both cases that inside template HTML files the URLs are with the slug.
### Renaming the files
Currently, some folder trees are only copied (assets), some are built with external tools and then copied (third_party) and some are built and then copied (extensions, templates). In this situation, the renaming would happen in multiple places in the code. I think that it would be better to rename everything in one place. 
So, we would need to design a better process for that. I think something like:
1. generate hashes for all files
2. build files that need to be build
3. copy all files to appropriate folders (here will the renaming happen)

One more difference between dev and prod would be only that the files are renamed.

One problem here are the directives, they would need to use slug too, but when we build, we only see HTML and don’t know if inside is directive or template (_base.html_ etc.)

**Upsides:**
- the build process would be refactored
- more future-proof

**Downsides:**
- the difference between dev and prod is little bit bigger
- the dev time would be longer
- problem with directives would be harder to solve
### Modifying the handler
We would need to change this handler:
```yaml
- url: /build
  static_dir: build
```
To something like this (in this example the filepath is `<path>/<name>.<slug>.<extension>`)
```yaml
- url: /build/(group 1)(group 2)(group 3)$
  static_files: build/\1\3
  upload: build/(group 1)(group 3)$
```
These are the regex capture groups.
_group 1_ would capture the part before slug
_group 2_ would capture the slug
_group 3_ would capture the part after slug

The line `static_files: build/\1\3` would remove the hash part. For example, request for file `build/extensions/something.d4sfdsd41f.js` would be looking for file `build/extensions/something.js`

There would be some tricky files like `third_party.min.<slug>.js` that we would also need to process by the regex. Also, in future, if we would add some file with an obscure name the regex could break.

Currently, we use the aforementioned handler for serving most of the files in prod and then other handlers to serve files in dev, so there would be no change in differences between dev and prod.

**Upsides:**
- there would be no added difference between dev and prod

**Downsides:**
- some files can break the regex
- the regex can be pretty complicated

BenHenning commented:
@vojtechjelinek Thanks for compiling the info! @seanlip and I discussed this a bit today. I may be wrong, but it seems like a lot of the work needed for the per-file approach would also be needed for the handler (ie, updating all JS/HTML file references to use the hash in frontend files). I think there is another downside unmentioned for the handler approach: it's not very intuitive, which may cause problems when debugging or maintaining this code in the future. Since it's conceptually more complicated and seems to require a lot of the same work as the per-file approach, it seems like the per-file approach may be ideal since it is more maintainable in the long-run. What do you think? Are we maybe missing something that may make the handler approach a much bigger win over the per-file approach?

vojtechjelinek commented:
Yes, you’re right that the handler approach is unintuitive and that the file renaming approach is safer and more maintainable and I can’t think of any big advantage for the handler approach.

seanlip commented:
OK! Let's go with the file renaming approach, then. Could you make the necessary changes to this PR? We can review again after that's done.

vojtechjelinek commented:
Yes, I will make the changes on this PR.

BenHenning commented:
Great, thanks @vojtechjelinek!

vojtechjelinek commented:
@BenHenning Hi, don't you know if there is any special reason why in production we use _oppia.css_ in _/build/css_ and not the _oppia.css_ in _/build/templates/head/css_? I tried to go through the commit history, but didn't find any info. It would simplify the build process a little bit.

seanlip commented:
@vojtechjelinek I don't think there's a good reason. I know there's a /css handler in app.yaml that aliases to core/templates/dev/head/css but I think that might be a legacy thing. If everything continues to work after the change then I think simplifying it is fine.

vojtechjelinek committed:
cache slug system changed

vojtechjelinek commented:
@BenHenning Could you please take a look at the general structure. 
The files are not renamed in /assets, because I would need to solve loading of images and translates there. Also, directives aren’t solved yet.

BenHenning commented:
Thanks @vojtechjelinek! I think this initially looks good, but I have an initial question: why can't we rename all files? At the moment, you're renaming the extension in the form:

_[filename].[extension]_ -> _[filename].[hash][extension]_

Is there a method of renaming we could use to preserve the extension?

I didn't look through the PR to make suggestions on improving the code, yet. I'm waiting to see how you solve some of the other parts you mentioned earlier, since I get the sense a lot more will need to be done to get that to work.

Regarding assets, do we need to still use a per-build slug system for those? If we want to turn on no-expire caching for all of Oppia's files, we need to make sure that some files (such as assets) are properly reloaded for new releases. What is your plan for those files?

vojtechjelinek commented:
I’m not sure what you mean. I plan to rename most of the files that are served, except for templates HTML (like _base.html_) because these are used only by the backend.

I'm renaming [filename].[extension] -> [filename].[hash]**.**[extension] so the extension is preserved. 

The assets still need slug system, especially translates, images — but I think it would be best to interpolate the hashes in the frontend. So create some JSON with the hashes of files in _/i18n_ and _/images_, for translates create a custom loader that will build the URL from hash and filename and for _/images_ modify _UrlInterpolationService_ to do the same.

BenHenning commented:
Ah, what confused me was ``.`` is included in the extension when you perform the concatenation:

``filepath_with_hash = file_name + '.' + file_hash + file_extension``

That being said, I'm a bit confused on not including hashes in HTML files. We have two different notions of templates: Angular templates and Jinja templates. Some of our HTML files are Angular templates, others are Jinja. We're generally moving away from using Jinja templates where we can. So to be clear, you intend not to rename HTML files that are Jinja templates, but do intend to rename directives & controller HTML files, is that right?

vojtechjelinek commented:
Yes, I intend not to rename only Jinja templates. Just need to find a way to differentiate between Jinja templates and other HTML files.

BenHenning commented:
Thanks @vojtechjelinek, is there anything else you'd like my thoughts on? :)

vojtechjelinek committed:
interpolationg of images, directives, translates

vojtechjelinek committed:
Subject images proper load

vojtechjelinek committed:
use Jinja for css url interpolation and use athor function for loading overlays

vojtechjelinek committed:
prod loading

vojtechjelinek committed:
fix tests

vojtechjelinek committed:
linting fixes

seanlip commented:
@vojtechjelinek -- just FYI, there are merge conflicts!

vojtechjelinek committed:
merge

seanlip commented:
Hi @vojtechjelinek, qq: is this ready for review, or is there more you want to do?

It might be a good idea to write up a checklist of remaining tasks here (see e.g. what I did in #3590) so that everyone's on the same page and reviewers have the appropriate context.

Thanks!

vojtechjelinek committed:
removed print

vojtechjelinek commented:
Everything is completed. Just select2 library is producing an error. I will take a look tomorrow.

seanlip commented:
Sounds good, thanks @vojtechjelinek :) In that case, please let us know once you're ready for a review!

vojtechjelinek commented:
@seanlip Hi, the issue with select2 is that the _select2.css_ is trying to load images that I have renamed (they include the hash in their name). To solve this I would need to create a list with files that shouldn't be renamed but I don't feel good about this because there is already enough exceptions in the build process. 

I propose switching to the newer version of the select2 library that doesn't use images. I tried to use the newer version but some other problems emerged so I think, if we switched, it would be better to do it in another pull request. What do you think?

seanlip commented:
Yep, I think you are making the right judgment calls here. Could you
upgrade select2 in another PR and get that merged into develop? Then we can
merge develop into this PR and complete it.

Thanks!

On Monday, July 3, 2017, Vojtěch Jelínek <notifications@github.com> wrote:

> @seanlip <https://github.com/seanlip> Hi, the issue with select2 is that
> the *select2.css* is trying to load images that I have renamed (they
> include the hash in their name). To solve this I would need to create a
> list with files that shouldn't be renamed but I don't feel good about this
> because there is already enough exceptions in the build process.
>
> I propose switching to the newer version of the select2 library that
> doesn't use images. I tried to use the newer version but some other
> problems emerged so I think, if we switched, it would be better to do it in
> another pull request. What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/3454#issuecomment-312700869>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeyv_4SdgIpiqy4yEAwdlM7-6posa1ks5sKSZvgaJpZM4NippJ>
> .
>

vojtechjelinek committed:
renamed constatant

vojtechjelinek commented:
@BenHenning Hi, you can start with the review.

seanlip commented on app.yaml:
```diff
@@ -32,6 +32,7 @@ handlers:
 - url: /build
   static_dir: build
   secure: always
+  application_readable: true
```
Why does /build need to be application_readable (by the server)?

vojtechjelinek replied:
The _application_readable_ is needed for the gadgets/RTEs/interactions _html_body_ function to work, because this function loads HTMLs in backend.

seanlip replied:
See other comment -- I wonder if we can do gadgets/etc. in a separate pass, given that there seems to be some additional work needed before we can do it "properly".

seanlip commented on app.yaml:
```diff
@@ -48,8 +49,7 @@ handlers:
     Cache-Control: 'public, max-age=60'
-- url: /css
-  # NB: not minified. TODO(sll): fix.
+- url: /templates/dev/head/css
```
Do we even need this stanza -- is it already captured below?

vojtechjelinek replied:
The rule below is for js files, but it would be probably better to extend it to capture CSS files too then to have two separate rules.

seanlip replied:
Sounds good, let's do that then. Thanks!

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -32,11 +32,26 @@ oppia.factory('UrlInterpolationService', [
       return GLOBALS.ASSET_DIR_PREFIX + resourcePath;
     };
 
+    var getUrlWithSlug = function(resourcePath) {
```
Can you add a docstring to this and getPrefixedUrl above explaining what each does? Lots of different concepts here.

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -32,11 +32,26 @@ oppia.factory('UrlInterpolationService', [
 
+    var getUrlWithSlug = function(resourcePath) {
+      validateResourcePath(resourcePath);
+      stripedPath = resourcePath.substring(1);
```
strippedPath (two 'p's) or relativeResourcePath, I think. Do you want to verify that the path starts with '/'?

seanlip replied:
Also, missing var?

vojtechjelinek replied:
I have modified the code in build.py to add '/' in front of the object keys so that this function was simplified a little bit.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -32,11 +32,26 @@ oppia.factory('UrlInterpolationService', [
+      validateResourcePath(resourcePath);
+      stripedPath = resourcePath.substring(1);
+      if (GLOBALS.MINIFICATION || !GLOBALS.DEV_MODE) {
+        if (!angular.isUndefined(hashes[stripedPath])) {
```
Would `if (hashes[strippedPath])` do?

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -32,11 +32,26 @@ oppia.factory('UrlInterpolationService', [
+      if (GLOBALS.MINIFICATION || !GLOBALS.DEV_MODE) {
+        if (!angular.isUndefined(hashes[stripedPath])) {
+          hash = hashes[stripedPath];
+          splitedPath = resourcePath.split('.');
```
I think this might be clearer if you do a find-from-right, since you really only want the last segment. (Just in case we end up in the future with files that already have more than one dot in them.)

vojtechjelinek replied:
This actually works as a find-from-right, it splits the path by '.' then adds the hash as a last but one and joins as a string with '.' again.

seanlip replied:
I know, but this isn't actually obvious -- the reader has to think a bit about what's going on.

Hence the suggestion to do the find-from-right a little more explicitly using e.g. lastIndexOf().

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -32,11 +32,26 @@ oppia.factory('UrlInterpolationService', [
+          hash = hashes[stripedPath];
+          splitedPath = resourcePath.split('.');
+          splitedPath.splice(splitedPath.length - 1, 0, hash);
+          resourcePath = splitedPath.join('.');
```
Try to avoid redefining variables -- it's confusing if the same var has different meanings in different parts of the code. Here, you could just return the result; no need to overwrite resourcePath.

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -32,11 +32,26 @@ oppia.factory('UrlInterpolationService', [
+          return resourcePath;
+        }
+      }
+      return resourcePath;
```
In this case, character 0 of the input string is included, and in the above case, character 0 is excluded (due to the substring(1)). I'm not sure I understand why?

vojtechjelinek replied:
The 0 was excluded only for hashes indexing, the returned string contained the 0 character.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -32,11 +32,26 @@ oppia.factory('UrlInterpolationService', [
       return GLOBALS.ASSET_DIR_PREFIX + resourcePath;
     };
 
+    var getUrlWithSlug = function(resourcePath) {
```
Please also add tests for newly-introduced functions.

vojtechjelinek replied:
In order to test these 'private' functions, I would need to return them (like I did for _getExtensionResourceUrl_) is this okay? (I would prefix them with '_')

seanlip replied:
Ah yes, good point. I think this function is complex enough that it should probably be tested, so yes, I think that what you suggest is fine. Thanks for checking!

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -130,56 +145,66 @@ oppia.factory('UrlInterpolationService', [
       getStaticImageUrl: function(imagePath) {
         validateResourcePath(imagePath);
-        return getCachePrefixedUrl('/assets/images' + imagePath);
+        return getPrefixedUrl(
```
This is a bit icky (the nested functions). If you can pull it off, it might be cleaner to combine this into a general "naive backend path --> fully-prefixed url with cache slug" function so that all the URL-munging happens cleanly in one place -- and let's add lots of tests for that one place to make absolutely sure everything behaves correctly.

seanlip replied:
Hi @vojtechjelinek, flagging this comment -- did you miss it?

vojtechjelinek replied:
Yeah, sorry I have missed it.
you mean instead of `getPrefixedUrl('/assets' + getUrlWithSlug(jsonPath));` something like `getCompleteUrl(prefix, path)`?

seanlip replied:
yep!

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -130,56 +145,66 @@ oppia.factory('UrlInterpolationService', [
         }
-        return getCachePrefixedUrl('/extensions/interactions/' +
-          interactionId + '/static/' + interactionId + '.png');
+        return getPrefixedUrl('/extensions' + getUrlWithSlug('/interactions/' +
```
Perhaps this should delegate to getExtensionResourceUrl().

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -130,56 +145,66 @@ oppia.factory('UrlInterpolationService', [
+       * returns the complete url path to that resource.
+       */
+      getExtensionResourceUrl: function(resourcePath) {
+        validateResourcePath(resourcePath);
```
Here and in other places: I notice we validate the resource path in getPrefixedUrl(). Do we actually need to do this validation twice?

vojtechjelinek replied:
No, we don't. Done.

seanlip commented on core/templates/dev/head/i18n.js:
```diff
@@ -37,6 +37,27 @@ oppia.constant('DEFAULT_TRANSLATIONS', {
   I18N_SIGNUP_LOADING: 'Loading'
 });
 
+oppia.factory('customLoader', [
```
Current convention (not fully implemented yet) is to use UpperCamelCase for services.

Also if this is only meant to be used by i18n then it's name should have i18n in it. Currently the name is rather vague.

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/i18n.js:
```diff
@@ -37,6 +37,27 @@ oppia.constant('DEFAULT_TRANSLATIONS', {
+oppia.factory('customLoader', [
+  '$http', '$q', 'UrlInterpolationService',
+  function ($http, $q, UrlInterpolationService) {
+    return function (options) {
```
nit: remove space before '('

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/i18n.js:
```diff
@@ -37,6 +37,27 @@ oppia.constant('DEFAULT_TRANSLATIONS', {
+    return function (options) {
+      var fileUrl = [
+        options.prefix,
+        options.key,
```
I don't see a key in what's passed into this...

vojtechjelinek replied:
Key is added later by i18n.

seanlip replied:
Hm ok. Might be worth adding a comment to explain.

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/i18n.js:
```diff
@@ -37,6 +37,27 @@ oppia.constant('DEFAULT_TRANSLATIONS', {
+        options.key,
+        options.suffix
+      ].join('');
+      return $http({
```
Could you use the usual $http.get?

vojtechjelinek replied:
Done.

seanlip commented on core/templates/dev/head/i18n.js:
```diff
@@ -37,6 +37,27 @@ oppia.constant('DEFAULT_TRANSLATIONS', {
+oppia.factory('customLoader', [
+  '$http', '$q', 'UrlInterpolationService',
+  function ($http, $q, UrlInterpolationService) {
+    return function (options) {
```
Is it possible for this to be function(prefix, key, suffix)? No worries if not, but if so maybe add some documentation for what `options` is expected to be.

vojtechjelinek replied:
No, because the function is called internally by i18n library.

seanlip commented on core/templates/dev/head/i18n.js:
```diff
@@ -37,6 +37,27 @@ oppia.constant('DEFAULT_TRANSLATIONS', {
   I18N_SIGNUP_LOADING: 'Loading'
 });
 
+oppia.factory('customLoader', [
```
Btw -- this is meant to be a service, right? It kind of looks more like a filter to me, so I'm mildly confused about what it's supposed to be. Which means that some of my other comments may be wrong.

vojtechjelinek replied:
This is service used by i18n. The function inside is called internally after the language that should be used is known.

seanlip commented on extensions/gadgets/base.py:
```diff
@@ -84,7 +84,8 @@ def html_body(self):
         """
         html_templates = utils.get_file_contents(os.path.join(
-            feconf.GADGETS_DIR, self.type, '%s.html' % self.type))
+            utils.get_asset_dir_prefix()[1:], feconf.GADGETS_DIR,
```
The 1: is a bit messy.

I wonder whether something different is needed here (and for interactions etc.). Couldn't we refer to the files statically somehow from the frontend? If so, and we still need to build that out, would it be worth leaving this as it used to be? I'm not sure I understand what this change buys us.

vojtechjelinek replied:
Previously the gadgets were loaded were loaded from same folder (oppia/extensions/gadegets/…) in dev mode and in prod mode too. Now, because the HTMLs are modified in prod mode (the URLs are replaced with URLs with hashes), we need to load them from build folder. 

The gadgets should probably be served statically, also the #2575 probably should have been solved before this.

(all stated for gadgets apply to RTCs and interactions too)

seanlip replied:
Yeah, I worry a bit about this, especially since this is causing changes to app.yaml as well. Do you think we could do the file hashes for everything except the gadgets/etc. in this PR, and leave the processing of extensions-html stuff for later until #2575 is handled?

vojtechjelinek replied:
In order to leave gadgets/etc. from this pull requests, I would again need to introduce the folder that would change the name in every new build (so that gadgets files would break the cache)and then the whole system would be useless (because all files would have new location for every new build). 

But, I have thought about it a little bit yesterday and I have changed the system a little bit so that in backend the files are loaded from _build_staging_ (here the files are build but not renamed), it also simplified few things in _build.py_

I will try to commit in few hours or tomorrow.

seanlip commented on scripts/build.py:
```diff
@@ -19,17 +19,26 @@
 import sys
-import yaml
-
+import hashlib
```
Alphabetize stdlib imports

vojtechjelinek replied:
Done.

seanlip commented on utils.py:
```diff
@@ -438,8 +438,8 @@ def get_thumbnail_icon_url_for_category(category):
-    return ('%s/assets/images/subjects/%s.svg'
-            % (get_asset_dir_prefix(), icon_name.replace(' ', '')))
+    return ('/subjects/%s.svg'
+            % (icon_name.replace(' ', '')))
```
Maybe move to previous line and drop outer parens?

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -19,17 +19,26 @@
     '..', 'oppia_tools', 'yuicompressor-2.4.8', 'yuicompressor-2.4.8.jar')
 
-FILE_EXTENSIONS_TO_IGNORE = ['.py']
+FILE_EXTENSIONS_TO_IGNORE = ('.py')
```
Without context, these constants are hard to understand. Could you provide some comments explaining what these represent and what rule governs the inclusion, within these lists, of specific files? Devs should be able to understand, for example, why '.py' files are ignored and others aren't, and what "ignoring" means in this context. It is OK if your comments are verbose.

I am also a bit confused about why we choose to rename files or not based on their extension (rather than their location or the context of their use).

vojtechjelinek replied:
Done.

I initially thought the files that we can't rename would be spread through the folders, but as I look at it he only files that are in multiple folders are '.map' I will rework this to exclude from renaming  by folders/files locations, not file extensions.

seanlip commented on scripts/build.py:
```diff
@@ -44,9 +53,15 @@ def ensure_directory_exists(f):
 
 
-def process_html(filename, target):
+def process_html(filename, target, file_hashes):
```
I think we're getting to the point where these functions need docstrings in the same format as the other backend files (see #2394). Could you please add them?

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -44,9 +53,15 @@ def ensure_directory_exists(f):
     f = open(filename, 'r')
     content = f.read()
+    for filepath, file_hash in file_hashes.iteritems():
+        if filepath.endswith('.html'):
```
I'm not sure I understand why we skip html files. Could you explain?

If it's not obvious, maybe also add an inline comment to clarify.

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+
+            target_path = source_path
+            relative_filepath = os.path.relpath(source_path, source)
+            if (not relative_filepath.endswith(FILE_EXTENSIONS_NOT_TO_RENAME) or
```
I find this confusing; what if both conditions conflict? Is there a reason we can't just rename everything? What if new file types get introduced later on?

(Again, it might be worth adding comments to explain the answers to questions like these.)

vojtechjelinek replied:
Reworked. See other comments.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+            if (not relative_filepath.endswith(FILE_EXTENSIONS_NOT_TO_RENAME) or
+                    relative_filepath.endswith(FILE_ENDINGS_TO_RENAME)):
+                filepath, file_extension = os.path.splitext(source_path)
+                target_path = (filepath + '.' +
```
Looks like you do this sort of construction in multiple places; I suggest pulling out a private function to handle it.

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+                               file_hashes[relative_filepath]+
+                               file_extension)
+
+            target_path = target_path.replace(source, target)
```
Perhaps we should be a bit more careful here. The expected replacement is only the prefix, right? If so, can we figure out the relative filepath and add the target_path prefix only at the end (ignoring the source_path entirely)?

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+        str. Hexadecimal hash of specified file.
+    """
+    m = hashlib.md5()
+    with open(filepath, "rb") as f:
```
Nit: use single quotes (codebase convention)

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+    """
+    file_hashes = dict()
+
+    print 'Processing %s' % os.path.join(os.getcwd(), folder_path)
```
Can we be more specific: something like "Computing file hash for X"? Ditto below and elsewhere. This helps with debugging.

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
 
 
-def build_files(source, target):
+def generate_file_md5(filepath):
```
generate_md5_hash(filepath)? (Otherwise, sounds like you're generating an md5 file.)

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+    return file_hashes
+
+
+def filter_hashes(file_hashes):
```
Might as well combine the logic here with the os.walk() in get_file_hashes(), and expose a single function for this purpose. That way you also save time not computing hashes for things that matter.

vojtechjelinek replied:
Here we filter the hashes that should be provided to the frontend, but even if we don't need them in the frontend we need them for the build process.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+def filter_hashes(file_hashes):
+    filtered_hashes = dict()
+    for filepath, file_hash in file_hashes.iteritems():
+        for folder in FOLDERS_PROVIDED_TO_FRONTEND:
```
All these filters need more documentation. I'm also wondering if we can simplify them a little bit. It's not clear how all the different constants interplay with each other.

Ideally you want a constant/function that's something like is_file_cacheable(), and that function acts as the single source of truth.

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+    return filtered_hashes
+
+
+def save_hashes_as_json(filepath, file_hashes):
```
Add docstring. In particular, `filepath` here seems to mean something different from the way it's used elsewhere, so maybe `target_filepath` might be a better name. The docstring should also explain that the JSON file will not actually end up at the target filepath.

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +115,95 @@ def copy_files_source_to_target(source, target):
+
+    hashes_json = json.dumps(filtered_hashes)
+    with open(filepath, 'w') as f:
+        f.write("var hashes = " + hashes_json)
```
Use single quotes. Btw, do we need a semicolon at the end, as well as quotes around hashes_json?

It also strikes me that there doesn't seem to be a test for some of these functions in the build process, which are rather important. Can we add a build_test.py file to e.g. check that the correct JSON is generated? I realize that you don't really want to actually generate files from test runs but maybe we could have a separate testable function along the lines of get_hashes_json_file_contents().

vojtechjelinek replied:
What do you mean by ‘as well as quotes around hashes_json’? _hashes_ should be a js object, if there were quotes it would be a string.

seanlip replied:
Hm, this is part of f.write and json.dumps() -- isn't it supposed to be a string?

I.e. don't you want to write something like "var hashes = JSON.parse('blah-blah-blah');"  where the blah-blah-blah is a JSON string? I was asking about the single quotes around the 'blah-blah-blah' but maybe I'm misunderstanding something.

vojtechjelinek replied:
Oh, I see. In my code, I add the Python generated JSON directly in JS so the generated file looks like:
```
var hashes = {'blah': 'blah'};
```
but the way with `JSON.parse()` is probably safer. I will change this.

seanlip commented on scripts/build.py:
```diff
@@ -142,50 +239,43 @@ def build_files(source, target):
+    # Minify extension static resources, create hashes for them and copy them
+    # into build/extensions
+    EXTENSIONS_DEV_DIR = os.path.join('extensions', '')
+    EXTENSIONS_PROD_DIR = os.path.join('extensions', 'prod', '')
```
I don't understand, why do we have an EXTENSIONS_PROD_DIR? Is this a temporary dir for staging?

If so, can we have a /build_staging dir where we do all this stuff so that we don't have random prod dirs all over the codebase?

vojtechjelinek replied:
Done. Also, since we now load some of the files from this folder in the backend (gadgets/etc.) some other name would probably be better.

seanlip commented on scripts/build.py:
```diff
@@ -142,50 +239,43 @@ def build_files(source, target):
+    TEMPLATES_OUT_DIR = os.path.join('build', 'templates', 'head', '')
+    hashes.update(get_file_hashes(TEMPLATES_DEV_DIR))
+
+    HASHES_JSON = os.path.join('build', 'assets', 'hashes.js')
```
hashes_json_filepath (this doesn't look like a constant)

seanlip commented on scripts/build.py:
```diff
@@ -142,50 +239,43 @@ def build_files(source, target):
+
+    HASHES_JSON = os.path.join('build', 'assets', 'hashes.js')
+    save_hashes_as_json(HASHES_JSON, hashes)
+    build_files(TEMPLATES_DEV_DIR, TEMPLATES_PROD_DIR, hashes)
```
Why do we separate these two lines from all the other templates stuff above?

vojtechjelinek replied:
The code changed a little bit, there is a comment now.

seanlip reviewed (changes requested):
Hi @vojtechjelinek -- Ben will only be able to take a pass at this on Monday, and he is actually going to be away starting mid-next-week. I think, given the time constraints, it'd be good if he could look at something that's almost final, so I've taken a preliminary pass. If you could get a new draft for review to Ben by Monday, that would be great; in the meantime, please let me know if you have questions!

vojtechjelinek committed:
part of comments addressed

vojtechjelinek committed:
changed extension loading, adressed comments

vojtechjelinek commented:
I have addressed some of the comments. I will try to finish rest of the things tomorow.

Things missing:
- [x] tests for _UrlInterpolationService.js_
- [x] tests for _build.py_
- [x] docstrings for _build.py_
- [x] extensions filtering → folder/file location filtering in _build.py_

seanlip commented on feconf.py:
```diff
@@ -51,9 +51,13 @@
-GADGETS_DIR = os.path.join('extensions', 'gadgets')
-RTE_EXTENSIONS_DIR = os.path.join('extensions', 'rich_text_components')
+
+EXTENSIONS_DIR_PREFIX = 'build_staging' if (IS_MINIFIED or not DEV_MODE) else ''
```
Hm, the extensions dir is really 'extensions' (or os.path.join('build_staging', 'extensions')). Maybe use that instead?

seanlip replied:
Also I think I'm slightly confused. Is the plan to upload build_staging/ to production and serve from there? My conception was that it is meant to be a temporary directory of sorts that's used only within the build process, but if so it shouldn't be given this sort of status within feconf. The prepending of build_staging should only happen in build.py.

vojtechjelinek replied:
I borrowed the concept from _templates/dev/head_ and _templates/prod/head_ because the files are built into the _prod_ folder and then copied into the _build_ folder. The backend uses HTML files from _prod_ and frontend from _build_. I do the same but with extensions. If used this way the name _build_staging_ is probably not appropriate.

seanlip replied:
Ah, I see.

Would you mind elaborating a bit on "the backend uses HTML files from _prod_"? I am wondering to what extent that is actually necessary. If we could transition (perhaps gradually) towards _prod_ not being necessary at all during runtime, that might be an improvement.

In the meantime, one idea is to remove core/templates/prod/head and have that in /build_staging/core/templates/dev/head (or something like that) so that all the prod files are in one place. That way we can say: here's the special "prod" folder, here's the special "build" folder, and everything else is dev. It would make the build process cleaner. Then over time perhaps we can get rid of prod altogether as we try to make the backend less dependent on files which should really be static anyway.

vojtechjelinek replied:
When backend loads HTML files to be parsed by Jinja and then served they are loaded from core/templates/prod/head, the only reason before this pull request was because they were minified there. This pull request introduces the second reason that is the paths inside the HTML are interpolated with hashes in the prod folder.

Moving that folder is good idea.

seanlip replied:
OK, thanks! Let's do that then, and see if we can reduce the prod footprint over time. We really need to document all this (i.e. the definitions of the folders) somewhere, too -- perhaps in the build.py script?

vojtechjelinek replied:
Should we also change the name of the folder? Because _build_staging_ could be a little bit misleading.

seanlip replied:
Yes, I think so. prod_files_for_backend/, maybe? Your call.

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -14,22 +14,38 @@
+FILE_EXTENSIONS_TO_IGNORE = ('.py')
+# Files with this extension should be moved to build folder but shouldn't be
+# renamed (i.e. the filepath shouldn't contain hash)
+# This is because these files are referenced from places we can't influence.
```
Can you elaborate briefly here? This sounds a bit cryptic.

Also, where exactly do these files come from? I don't remember having .eot, .woff2, etc. files in our codebase, so I'm a bit confused.

vojtechjelinek replied:
Reworked to filter by paths not by extensions.

seanlip commented on scripts/build.py:
```diff
@@ -14,22 +14,38 @@
+
+# Hashes for files in these folders should be provided to frontend in
+# JS hashes object.
+FOLDERS_PROVIDED_TO_FRONTEND = (os.path.join('images', ''),
```
What do you think about specifying both this and the next constant as a single set of glob patterns instead -- something like FILEPATHS_TO_HASH? We do something like this in scripts/pre_commit_linter.py and conceptually it might be more straightforward. Ideally we get it to the point where we can just edit that single list and all the hashing gets taken care of transparently.

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -14,22 +14,38 @@
+                                os.path.join('i18n', ''))
+# Hashes for files ending with these strings should be provided to frontend in
+# JS hashes object.
+FILE_ENDINGS_PROVIDED_TO_FRONTEND = ('directive.html',
```
Maybe _directive.html, _modal.html just to be a bit more strict.

What is overlay.html? Not familiar with that one -- I don't think it's one of our frontend conventions, and I don't think it should be given special treatment.

vojtechjelinek replied:
overlay.html is used in extensions/gadgets/AdviceBar/ and extensions/gadgets/TestGadget

seanlip replied:
Ah, OK. Can it be excluded from this list? It's main status is really just "part of the static files needed for a gadget"; the suffix "overlay.html" does not actually convey anything in particular.

We could also just rename the relevant files to end with _modal.html.

vojtechjelinek replied:
I renamed the files. Done.

seanlip commented on scripts/build.py:
```diff
@@ -100,13 +125,107 @@ def copy_files_source_to_target(source, target):
+
+    hashes_json = json.dumps(filtered_hashes)
+    with open(target_filepath, 'w') as f:
+        f.write('var hashes = ' + hashes_json + ';')
```
Generally we do interpolation: `'var hashes = %s;' % (...)`

Also, flagging that this needs JSON.parse().

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -142,50 +257,49 @@ def build_files(source, target):
-    ASSETS_SRC_DIR = os.path.join('assets', '')
-    ASSETS_OUT_DIR = os.path.join(BUILD_DIR, 'assets', '')
-    copy_files_source_to_target(ASSETS_SRC_DIR, ASSETS_OUT_DIR)
+    hashes = dict() # pylint: disable=C0103
```
pragmas should use the human-readable names, not codes.

But I don't understand. Why is the name not OK? What does pylint want it to be?

vojtechjelinek replied:
PyLint says 'Invalid constant name "hashes"' which is weird because I don't know how PyLint deduced that this is constant.

seanlip replied:
Ah, I think I know. It might be because it's in global scope. To get around this, declare a new function and have the bottom of the file just be something like

```
    if __name__ == '__main__':
        build_files()   <-- pick a suitable name here based on whatever the function does
```

I think that might fix it.

vojtechjelinek replied:
Done.

vojtechjelinek committed:
next batch of comments adressed

vojtechjelinek committed:
fix tests

vojtechjelinek committed:
UrlInterpolation tests

seanlip commented:
Hi @vojtechjelinek -- @BenHenning had less time than he expected to look at this because a couple of things came up this week before his trip, but he did take a quick look at it this evening and here's what he told me: "This is really cool. Sorry I didn't look at it earlier, but it generally looks good to me. Couple of small PR nits I could mention, but at a high-level it's really good, I think."

So, I think that's good! :) I will take over (or rather complete) the full review from here, since @BenHenning won't be around for a bit. Would you mind finishing up the last tests and addressing everything else? Then I'll do another pass and we should be able to get this in.

Thanks!

vojtechjelinek committed:
build tests and test fixes

vojtechjelinek commented:
Nice, I am glad.

I finished the backend tests and I think all comments are addressed. You can take a look.

seanlip commented:
Thanks, @vojtechjelinek! Just a quick question -- I removed the "[not for merge]" part of the title (since this is presumably almost ready for merge?) but I'm not sure in what sense this is "part of #2447". What is the other part?

vojtechjelinek commented:
I supposed that I would split this into parts, but it wasn't possible to split it. I changed the title.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationServiceSpec.js:
```diff
@@ -25,6 +25,38 @@ describe('URL Interpolation Service', function() {
+  });
+
+  it('should build complete URL with prefixes and hash', function () {
+    expect(uis._getCompleteUrl('/test_folder','/hash_test.html')).toBe(
```
Nit: missing space after comma, here and below.

vojtechjelinek replied:
Done.

seanlip reviewed (approved):
I've read the whole thing, and I can't see any further issues (save a small nit). This is fantastic, thank you @vojtechjelinek!

Also, one question. Could you please write up a checklist of what steps we should take to get to a build process that's ultimately as simple/clean and easy to understand and maintain as possible? I know there are a few things we want to improve, like our handing of gadgets/interactions/etc., but I'm not sure what other things need to be done and am missing the big picture a bit. It would be nice to have some more clarity on that.

Thanks!

seanlip commented:
Btw, there seems to be a linter failure in Travis.

vojtechjelinek committed:
fix linting errors

vojtechjelinek committed:
addressed comment

vojtechjelinek commented:
I will write up the checklist tomorrow. Should I create a new issue with that info?

seanlip commented:
Yes, please -- that sounds good. Thanks @vojtechjelinek!

vojtechjelinek commented:
(just to remind myself) remove /build folder when starting build process

seanlip commented:
NB per discussion with @vojtechjelinek, we should **hold off on merging this** until the following are done:

- [ ] Remove /build folder when starting build process
- [x] Remove cssBackgroundImage from dependencies system (since the images referenced from third party css are problematic for new cache slug system)

vojtechjelinek committed:
fix tests and remove cssBackgroundImages

vojtechjelinek commented:
@seanlip The 'Remove /build folder when starting build process' isn't needed because we already clean the folders in `copy_files_source_to_targe` function.

seanlip commented:
OK, thanks. Shall we merge this, then? I have no objections so please feel free to do so if you are happy with it!
